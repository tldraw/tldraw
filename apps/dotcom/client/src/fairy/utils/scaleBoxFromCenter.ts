import { BoxLike } from 'tldraw'

export function scaleBoxFromCenter(box: BoxLike, n: number): BoxLike {
	const centerX = box.x + box.w / 2
	const centerY = box.y + box.h / 2
	
	const newW = box.w * n
	const newH = box.h * n
	
	return {
		x: centerX - newW / 2,
		y: centerY - newH / 2,
		w: newW,
		h: newH,
	}
}

