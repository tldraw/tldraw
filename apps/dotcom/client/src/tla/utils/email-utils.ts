// Type definitions for better type safety
export interface UserForUsernamePicker {
	fullName?: string | null
	emailAddresses: Array<{ emailAddress: string }>
}

/**
 * Extract and format username from email address
 * @param email - The email address to extract from
 * @returns Formatted display name or fallback
 */
export function extractUsernameFromEmail(email: string): string {
	if (!email?.includes('@')) return 'user'

	const username = email.split('@')[0]
	if (username.length < 2) return 'user'

	// Clean and format username in one operation
	return (
		username
			.replace(/[._+-]+/g, ' ')
			.replace(/\s+/g, ' ')
			.trim()
			.toLowerCase()
			.split(' ')
			.map((word) => word.charAt(0).toUpperCase() + word.slice(1))
			.join(' ') || 'user'
	)
}

/**
 * Check if user needs username picker
 * @param user - The user object
 * @returns True if user needs to pick a username
 */
export function needsUsernamePicker(user: UserForUsernamePicker): boolean {
	// Early return if user already has a name
	if (user.fullName?.trim()) return false

	// Check if user has a valid email address
	return Boolean(user.emailAddresses?.[0]?.emailAddress?.includes('@'))
}
