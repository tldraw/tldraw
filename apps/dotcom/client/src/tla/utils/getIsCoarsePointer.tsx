import { tlenv } from 'tldraw'

const mql =
	(typeof window !== 'undefined' &&
		window.matchMedia &&
		window.matchMedia('(any-pointer: coarse)')) ||
	undefined

// This is a workaround for a Firefox bug where we don't correctly
// detect coarse VS fine pointer. For now, let's assume that you have a fine
// pointer if you're on Firefox on desktop.
const isForcedFinePointer = tlenv.isFirefox && !tlenv.isAndroid && !tlenv.isIos

export function getIsCoarsePointer() {
	// default to true if matchMedia is not supported, which is extremely unlikely
	return isForcedFinePointer ? false : mql?.matches ?? true
}
