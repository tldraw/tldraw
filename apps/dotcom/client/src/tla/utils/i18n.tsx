/* eslint-disable no-restricted-imports */

import {
	FormattedMessage,
	IntlConfig,
	IntlShape,
	MessageDescriptor,
	PrimitiveType,
	createIntlCache,
	createIntl as originalCreateIntl,
	defineMessages as originalDefineMessages,
	useIntl,
} from 'react-intl'

import { FormatXMLElementFn } from 'intl-messageformat'
import md5 from 'md5'
import { ComponentPropsWithoutRef } from 'react'

// Re-export everything and override below what we want to override.
// eslint-disable-next-line
export * from 'react-intl'

export function useMsg(
	message: MessageDescriptor,
	values?: Record<string, PrimitiveType | FormatXMLElementFn<string, string>>
) {
	const intl = useIntl()
	return intl.formatMessage(message, values)
}

const INTERNAL_LOCALES = ['xx-AE', 'xx-LS']

// This matches the extraction tool pattern:
//   --id-interpolation-pattern '[md5:contenthash:hex:10]'
function generateId({ id, description, defaultMessage }: MessageDescriptor) {
	if (id) {
		return id
	}

	return md5((description ? `${defaultMessage}#${description}` : defaultMessage) as string).slice(
		0,
		10
	)
}

export function F(props: ComponentPropsWithoutRef<typeof FormattedMessage>) {
	const intl = useIntl()
	const id = generateId(props)
	let internalMessage = (props.defaultMessage || '') as string
	if (intl.locale === 'xx-AE') {
		internalMessage = makeAccented(internalMessage)
	} else if (intl.locale === 'xx-LS') {
		internalMessage = makeLong(internalMessage)
	}

	return (
		<span className="i18n-msg">
			{isInternalLocale(intl.locale) ? (
				<>{internalMessage}</>
			) : (
				/* eslint-disable-next-line formatjs/enforce-default-message */
				<FormattedMessage id={id} {...props} />
			)}
		</span>
	)
}

// We programmatically define ID's for messages to make things easier for devs.
export function defineMessages<T extends string, D extends MessageDescriptor>(
	msgs: Record<T, D>
): Record<T, D> {
	for (const key in msgs) {
		if (!msgs[key].id) {
			msgs[key].id = generateId(msgs[key])
		}
	}
	return originalDefineMessages(msgs)
}

export function isInternalLocale(locale: string) {
	return INTERNAL_LOCALES.indexOf(locale) !== -1
}

function makeAccented(str: string) {
	const accents = 'áƃçđéƒǵȟíǰķłɱñóƥɋřšťúṽẃẍýž'
	return str.replace(/[a-zA-Z]/g, (char) => {
		const isUpper = char === char.toUpperCase()
		const index = char.toLowerCase().charCodeAt(0) - 97
		const accentedChar = accents[index] || char
		return isUpper ? accentedChar.toUpperCase() : accentedChar
	})
}

function makeLong(str: string) {
	return `${str} loooooooo oo ooooooong`
}

// This is optional but highly recommended since it prevents memory leaks.
// See: https://formatjs.github.io/docs/react-intl/api/#createintl
const cache = createIntlCache()
let presetIntl: IntlShape | null = null
let didSetupCreateIntl = false
export function setupCreateIntl({ defaultLocale, locale, messages }: IntlConfig) {
	presetIntl = originalCreateIntl(
		{
			defaultLocale,
			locale,
			messages,
		},
		cache
	)

	didSetupCreateIntl = true
}

// createIntl is used in non-React locations.
export function createIntl(options?: IntlShape) {
	if (options) {
		return originalCreateIntl(options)
	} else {
		if (!didSetupCreateIntl) {
			throw new Error('Need to run setupCreateIntl to use createIntl without options.')
		}
		return presetIntl
	}
}
