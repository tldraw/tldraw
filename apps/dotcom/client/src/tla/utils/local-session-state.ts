import { TlaUser } from '@tldraw/dotcom-shared'
import {
	atom,
	deleteFromLocalStorage,
	getFromLocalStorage,
	setInLocalStorage,
	useValue,
} from 'tldraw'

const STORAGE_KEY = 'tldrawapp_session_3'

export interface TldrawAppSessionState {
	createdAt: number
	isSidebarOpen: boolean
	isSidebarOpenMobile: boolean
	auth?: {
		userId: string
	}
	shareMenuActiveTab: 'share' | 'export' | 'publish' | 'anon-share'
	sidebarActiveTab: 'recent' | 'groups' | 'shared' | 'drafts' | 'starred'
	theme: 'light' | 'dark'
	views: {
		[key: string]: {
			sort: 'recent' | 'newest' | 'oldest' | 'atoz' | 'ztoa'
			view: 'grid' | 'list'
			search: string
		}
	}
	flags: { [key: string]: boolean }
	exportSettings: Pick<
		TlaUser,
		'exportFormat' | 'exportTheme' | 'exportBackground' | 'exportPadding'
	>
	sidebarWidth?: number
}

let prev: TldrawAppSessionState = {
	createdAt: Date.now(),
	isSidebarOpen: true,
	isSidebarOpenMobile: false,
	shareMenuActiveTab: 'share',
	sidebarActiveTab: 'recent',
	theme: 'light',
	views: {},
	flags: {},
	exportSettings: {
		exportFormat: 'png',
		exportTheme: 'auto',
		exportBackground: true,
		exportPadding: true,
	},
	sidebarWidth: 260,
}

try {
	const stored = getFromLocalStorage(STORAGE_KEY)
	const prevStored = stored ? JSON.parse(stored) : null
	if (prevStored) {
		prev = {
			...prev,
			...(prevStored as TldrawAppSessionState),
			// forced initial values
			createdAt: Date.now(),
			isSidebarOpenMobile: false,
			auth: undefined,
		}
	}
} catch {
	// noop
}

const localSessionState = atom<TldrawAppSessionState>('session', prev)

export function clearLocalSessionState() {
	return deleteFromLocalStorage(STORAGE_KEY)
}
export function getLocalSessionStateUnsafe() {
	return localSessionState.__unsafe__getWithoutCapture()
}

export function getLocalSessionState() {
	return localSessionState.get()
}

export function setLocalSessionState(state: TldrawAppSessionState) {
	localSessionState.set(state)
	setInLocalStorage(STORAGE_KEY, JSON.stringify(localSessionState.get()))
}

export function updateLocalSessionState(
	fn: (state: TldrawAppSessionState) => Partial<TldrawAppSessionState>
) {
	localSessionState.update((state) => {
		return { ...state, ...fn(state) }
	})
	setInLocalStorage(STORAGE_KEY, JSON.stringify(localSessionState.get()))
}

export function useLocalSessionState() {
	return useValue('session', () => getLocalSessionState(), [])
}
