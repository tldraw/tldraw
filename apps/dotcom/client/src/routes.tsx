import { captureException } from '@sentry/react'
import { TLRemoteSyncError, TLSyncErrorCloseEventReason } from '@tldraw/sync-core'
import { Suspense, lazy, useEffect } from 'react'
import { Helmet } from 'react-helmet-async'
import { Outlet, Route, createRoutesFromElements, useRouteError } from 'react-router-dom'
import { getFromLocalStorage } from 'tldraw'
import { DefaultErrorFallback } from './components/DefaultErrorFallback/DefaultErrorFallback'
import { ErrorPage } from './components/ErrorPage/ErrorPage'
import { notFound } from './pages/not-found'
import { ROUTES } from './routeDefs'
import { IntlProvider } from './tla/utils/i18n'
import { TlaNotFoundError } from './tla/utils/notFoundError'

const LoginRedirectPage = lazy(() => import('./components/LoginRedirectPage/LoginRedirectPage'))

const clerkCookieName = '__session'
export const tlaOverrideFlagName = 'tla-override-flag'

const isClerkCookieSet = document.cookie
	.split(';')
	.some((item) => item.trim().startsWith(clerkCookieName))

const isOverrideFlagSet = !!getFromLocalStorage(tlaOverrideFlagName) || navigator.webdriver

export const legacyRoutes = (
	<Route errorElement={<DefaultErrorFallback />}>
		<Route path={ROUTES.legacyRoot} lazy={() => import('./pages/root')} />
		{/* We don't want to index multiplayer rooms */}
		<Route element={<NoIndex />}>
			<Route lazy={() => import('./tla/providers/TlaRootProviders')}>
				<Route path={ROUTES.tlaOptIn} lazy={() => import('./pages/tla-opt-in')} />
				<Route path={ROUTES.tlaOverride} lazy={() => import('./pages/tla-override')} />
				<Route path={ROUTES.tlaFile} lazy={() => import('./tla/pages/file')} />
				<Route path={ROUTES.tlaPublish} lazy={() => import('./tla/pages/publish')} />
			</Route>
			<Route element={<ShimIntlProvider />}>
				<Route path={ROUTES.legacyNewPage} lazy={() => import('./pages/new')} />
				<Route path={ROUTES.legacyNewPage2} lazy={() => import('./pages/new')} />
				<Route
					path={ROUTES.touchscreenSidePanel}
					lazy={() => import('./pages/public-touchscreen-side-panel')}
				/>
				<Route path={ROUTES.legacyRoom} lazy={() => import('./pages/public-multiplayer')} />
				<Route path={ROUTES.legacyRoomHistory} lazy={() => import('./pages/history')} />
				<Route
					path={ROUTES.legacyRoomHistorySnapshot}
					lazy={() => import('./pages/history-snapshot')}
				/>
				<Route path={ROUTES.legacySnapshot} lazy={() => import('./pages/public-snapshot')} />
				<Route
					path={ROUTES.legacyReadonlyOld}
					lazy={() => import('./pages/public-readonly-legacy')}
				/>
				<Route path={ROUTES.legacyReadonly} lazy={() => import('./pages/public-readonly')} />
			</Route>
		</Route>
	</Route>
)

export const tlaRoutes = (
	<Route>
		<Route element={<ShimIntlProvider />}>
			<Route
				path={ROUTES.touchscreenSidePanel}
				lazy={() => import('./pages/public-touchscreen-side-panel')}
			/>
		</Route>
		<Route lazy={() => import('./tla/providers/TlaRootProviders')}>
			<Route path={ROUTES.tlaRoot} lazy={() => import('./tla/pages/local')} />
			<Route element={<NoIndex />}>
				<Route path={ROUTES.tlaLocalFile} lazy={() => import('./tla/pages/local-file')} />
				{/* <Route path={ROUTES.tlaPlayground} lazy={() => import('./tla/pages/playground')} /> */}
				{/* File view */}
				<Route path={ROUTES.tlaFile} lazy={() => import('./tla/pages/file')} />
				<Route path={ROUTES.tlaPublish} lazy={() => import('./tla/pages/publish')} />
				{/* Legacy room */}
				<Route path={ROUTES.tlaLegacyRoom} lazy={() => import('./tla/pages/legacy-room')} />
				{/* Legacy readonly */}
				<Route path={ROUTES.tlaLegacyReadonly} lazy={() => import('./tla/pages/legacy-readonly')} />
				<Route
					path={ROUTES.tlaLegacyReadonlyOld}
					lazy={() => import('./tla/pages/legacy-readonly-old')}
				/>
				{/* Legacy new */}
				<Route path={ROUTES.legacyNewPage} lazy={() => import('./pages/new')} />
				{/* Legacy snapshot */}
				<Route path={ROUTES.tlaLegacySnapshot} lazy={() => import('./tla/pages/legacy-snapshot')} />
				{/* Legacy history */}
				<Route
					path={ROUTES.tlaLegacyRoomHistory}
					lazy={() => import('./tla/pages/legacy-history')}
				/>
				{/* Legacy history snapshot */}
				<Route
					path={ROUTES.tlaLegacyRoomHistorySnapshot}
					lazy={() => import('./tla/pages/legacy-history-snapshot')}
				/>
				{/* Views that require login */}
				<Route lazy={() => import('./tla/providers/RequireSignedInUser')}></Route>
				<Route path={ROUTES.tlaOverride} lazy={() => import('./pages/tla-override')} />
			</Route>
		</Route>
	</Route>
)

export const router = createRoutesFromElements(
	<Route
		ErrorBoundary={() => {
			const error = useRouteError()
			useEffect(() => {
				captureException(error)
			}, [error])

			let header = 'Something went wrong'
			let para1 =
				'Please try refreshing the page. Still having trouble? Let us know at hello@tldraw.com.'
			if (error instanceof TLRemoteSyncError) {
				switch (error.reason) {
					case TLSyncErrorCloseEventReason.NOT_FOUND: {
						header = 'Not found'
						para1 = 'The file you are looking for does not exist.'
						break
					}
					case TLSyncErrorCloseEventReason.NOT_AUTHENTICATED: {
						return (
							<Suspense>
								<LoginRedirectPage />
							</Suspense>
						)
					}
					case TLSyncErrorCloseEventReason.FORBIDDEN: {
						header = 'Invite only'
						para1 = `You don't have permission to view this room.`
						break
					}
					case TLSyncErrorCloseEventReason.RATE_LIMITED: {
						header = 'Rate limited'
						para1 = `Please slow down.`
						break
					}
				}
			}
			if (error instanceof TlaNotFoundError) {
				return notFound()
			}

			return (
				<ErrorPage
					messages={{
						header,
						para1,
					}}
				/>
			)
		}}
	>
		{isClerkCookieSet || isOverrideFlagSet ? tlaRoutes : legacyRoutes}
		<Route path="/__debug-tail" lazy={() => import('./tla/pages/worker-debug-tail')} />
		<Route path="*" lazy={() => import('./pages/not-found')} />
	</Route>
)

function NoIndex() {
	return (
		<>
			<Helmet>
				<meta name="robots" content="noindex, noimageindex, nofollow" />
			</Helmet>
			<Outlet />
		</>
	)
}

function ShimIntlProvider() {
	return (
		// This IntlProvider is just for backwards compatibilty for the old site.
		<IntlProvider defaultLocale="en" locale="en" messages={{}}>
			<Outlet />
		</IntlProvider>
	)
}
