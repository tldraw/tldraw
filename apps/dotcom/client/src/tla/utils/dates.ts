import { FormatDateOptions } from '../utils/i18n'

export function getRelevantDates() {
	const now = new Date()

	return {
		today: new Date(now.getFullYear(), now.getMonth(), now.getDate()).getTime(),
		yesterday: new Date(now.getFullYear(), now.getMonth(), now.getDate() - 1).getTime(),
		thisWeek: new Date(now.getFullYear(), now.getMonth(), now.getDate() - now.getDay()).getTime(),
		thisMonth: new Date(now.getFullYear(), now.getMonth(), 1).getTime(),
		thisYear: new Date(now.getFullYear(), 0, 1).getTime(),
	}
}

export function getDateFormat(date: Date): FormatDateOptions {
	const { yesterday, thisMonth, thisYear } = getRelevantDates()
	const d = date.getTime()

	if (d >= yesterday) {
		return {
			weekday: 'short',
			hour: 'numeric',
			minute: 'numeric',
		}
	} else if (d >= thisMonth) {
		return {
			month: 'short',
			weekday: 'short',
			day: 'numeric',
			hour: 'numeric',
		}
	} else if (d >= thisYear) {
		return {
			month: 'short',
			weekday: 'short',
			day: 'numeric',
			hour: 'numeric',
		}
	}
	return {
		year: 'numeric',
		month: 'short',
		weekday: 'short',
		day: 'numeric',
	}
}
