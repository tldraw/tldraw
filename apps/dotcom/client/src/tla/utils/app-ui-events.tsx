import { TlaFile, TlaUser } from '@tldraw/dotcom-shared'
import { ReactNode, createContext, useContext } from 'react'
import { trackAnalyticsEvent } from '../../utils/trackAnalyticsEvent'
import { TldrawAppSessionState } from './local-session-state'

/** @public */
export type TLAppUiEventSource =
	| 'sidebar'
	| 'user-preferences'
	| 'file-rename-dialog'
	| 'file-menu'
	| 'file-share-menu'
	| 'file-header'
	| 'anon-landing-page'
	| 'anon-top-bar'
	| 'account-menu'
	| 'top-bar'
	| 'legacy-import-button'

/** @public */
export interface TLAppUiEventMap {
	'create-file': null
	'delete-file': null
	'rename-file': { name: string }
	'duplicate-file': null
	'drop-tldr-file': null
	'import-tldr-file': null
	'change-user-name': null
	'click-watermark': null
	'open-share-menu': null
	'change-share-menu-tab': { tab: TldrawAppSessionState['shareMenuActiveTab'] }
	'copy-share-link': null
	'copy-file-link': null
	'toggle-shared': { shared: boolean }
	'set-theme': { theme: 'dark' | 'light' | 'auto' }
	'toggle-export-padding': { padding: boolean }
	'toggle-export-background': { background: TlaUser['exportBackground'] }
	'set-export-format': { format: TlaUser['exportFormat'] }
	'set-export-theme': { theme: TlaUser['exportTheme'] }
	'export-image': {
		fullPage: boolean
		theme: TlaUser['exportTheme']
		format: TlaUser['exportFormat']
		padding: TlaUser['exportPadding']
		background: TlaUser['exportBackground']
	}
	'set-shared-link-type': { type: TlaFile['sharedLinkType'] | 'no-access' }
	'open-url': { url: string }
	'publish-file': null
	'unpublish-file': null
	'copy-publish-link': null
	'sign-in-clicked': null
	'sign-up-clicked': null
	'sign-out-clicked': null
	'learn-more-button': null
	'sidebar-toggle': { value: boolean }
	'click-file-link': null
}

/** @public */
export type TLAppUiHandler = <T extends keyof TLAppUiEventMap>(
	name: T,
	data: { source: TLAppUiEventSource } & (TLAppUiEventMap[T] extends null
		? object
		: TLAppUiEventMap[T])
) => void

/** @public */
export type TLAppUiContextType = TLAppUiHandler

/** @internal */
const defaultEventHandler: TLAppUiContextType = trackAnalyticsEvent

/** @internal */
export const EventsContext = createContext<TLAppUiContextType>(defaultEventHandler)

/** @public */
export interface TldrawAppUiEventsProviderProps {
	onEvent?: TLAppUiHandler
	children: ReactNode
}

/** @public @react */
export function TldrawAppUiEventsProvider({ onEvent, children }: TldrawAppUiEventsProviderProps) {
	return (
		<EventsContext.Provider value={onEvent ?? defaultEventHandler}>
			{children}
		</EventsContext.Provider>
	)
}

/** @public */
export function useTldrawAppUiEvents(): TLAppUiContextType {
	const eventHandler = useContext(EventsContext)
	return eventHandler
}
