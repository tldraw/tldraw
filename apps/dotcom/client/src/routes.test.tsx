import { expect } from '@jest/globals'
import { uniq } from '@tldraw/utils'
import type { MatcherFunction } from 'expect'
import { join } from 'path'
import { ReactElement } from 'react'
import { Route, RouteObject, createRoutesFromElements } from 'react-router-dom'
import { legacyRoutes, tlaRoutes } from './routes'

const toMatchAny: MatcherFunction<[regexes: unknown]> = function (actual, regexes) {
	if (
		typeof actual !== 'string' ||
		!Array.isArray(regexes) ||
		regexes.some((regex) => typeof regex !== 'string')
	) {
		throw new Error('toMatchAny takes a string and an array of strings')
	}

	const pass = regexes.map((regex) => new RegExp(regex)).some((regex) => actual.match(regex))
	if (pass) {
		return {
			message: () =>
				`expected ${this.utils.printReceived(actual)} not to match any of the regexes ${this.utils.printExpected(regexes)}`,
			pass: true,
		}
	} else {
		return {
			message: () =>
				`expected ${this.utils.printReceived(actual)} to match at least one of the regexes ${this.utils.printExpected(regexes)}`,
			pass: false,
		}
	}
}

expect.extend({ toMatchAny })

function extractContentPaths(routeObject: RouteObject): string[] {
	const path = routeObject.path || ''

	const paths: string[] = []
	if (
		path &&
		(routeObject.element || routeObject.Component || routeObject.lazy || routeObject.loader)
	) {
		// This is a contentful route so it gets included on its own.
		// Technically not every route with a .lazy or a .loader has content, but we don't have a way to check that
		// and it's not a huge deal if they don't 404 correctly.
		paths.push(path)
	}
	if (routeObject.children && routeObject.children.length > 0) {
		// this route has children, so we need to recurse
		paths.push(
			...routeObject.children.flatMap((child) =>
				extractContentPaths(child).map((childPath) => join(path, childPath))
			)
		)
	}

	return paths
}

function convertReactToVercel(path: string): string {
	// react-router supports wildcard routes https://reactrouter.com/en/main/route/route#splats
	// but we don't use them yet so just fail for now until we need them (if ever)
	if (path.endsWith('*')) {
		throw new Error(`Wildcard routes like '${path}' are not supported yet (you can add support!)`)
	}
	// react-router supports optional route segments https://reactrouter.com/en/main/route/route#optional-segments
	// but we don't use them yet so just fail for now until we need them (if ever)
	if (path.match(/\?\//)) {
		throw new Error(
			`Optional route segments like in '${path}' are not supported yet (you can add this)`
		)
	}
	// make sure colons are immediately preceded by a slash
	if (path.match(/[^/]:/)) {
		throw new Error(`Colons in route segments must be immediately preceded by a slash in '${path}'`)
	}
	if (!path.startsWith('/')) {
		throw new Error(`Route paths must start with a slash, but '${path}' does not`)
	}
	// Wrap in explicit start and end of string anchors (^ and $)
	// and replace :param with [^/]* to match any string of non-slash characters, including the empty string
	return '^' + path.replace(/:[^/]+/g, '[^/]*') + '/?$'
}

function extract(...routes: ReactElement[]) {
	return createRoutesFromElements(
		<Route>{routes.map((r, i) => ({ ...r, key: i.toString() }))}</Route>
	)
		.flatMap(extractContentPaths)
		.sort()
}

const spaRoutes = uniq(extract(legacyRoutes).concat(extract(tlaRoutes)))
	.sort()
	// ignore the root catch-all route
	.filter((path) => path !== '/*' && path !== '*')
	.map((path) => ({
		reactRouterPattern: path,
		vercelRouterPattern: convertReactToVercel(path),
	}))

const allvercelRouterPatterns = spaRoutes.map((route) => route.vercelRouterPattern)

test('the_routes', () => {
	expect(spaRoutes).toMatchSnapshot()
})

test('all React routes match', () => {
	for (const route of spaRoutes) {
		expect(route.reactRouterPattern).toMatch(new RegExp(route.vercelRouterPattern))
		for (const otherRoute of spaRoutes) {
			if (route === otherRoute) continue
			expect(route.reactRouterPattern).not.toMatch(new RegExp(otherRoute.vercelRouterPattern))
		}
	}
})

test("non-react routes don't match", () => {
	// lil smoke test for basic patterns
	expect('/').toMatchAny(allvercelRouterPatterns)
	expect('/new').toMatchAny(allvercelRouterPatterns)
	expect('/r/whatever').toMatchAny(allvercelRouterPatterns)
	expect('/r/whatever/').toMatchAny(allvercelRouterPatterns)

	expect('/assets/test.png').not.toMatchAny(allvercelRouterPatterns)
	expect('/twitter-social.png').not.toMatchAny(allvercelRouterPatterns)
	expect('/robots.txt').not.toMatchAny(allvercelRouterPatterns)
	expect('/static/css/index.css').not.toMatchAny(allvercelRouterPatterns)
})

test('convertReactToVercel', () => {
	expect(() =>
		convertReactToVercel('/r/:roomId/history?/:timestamp')
	).toThrowErrorMatchingInlineSnapshot(
		`"Optional route segments like in '/r/:roomId/history?/:timestamp' are not supported yet (you can add this)"`
	)
	expect(() => convertReactToVercel('/r/:roomId/history/*')).toThrowErrorMatchingInlineSnapshot(
		`"Wildcard routes like '/r/:roomId/history/*' are not supported yet (you can add support!)"`
	)
	expect(() => convertReactToVercel('/r/foo:roomId/history')).toThrowErrorMatchingInlineSnapshot(
		`"Colons in route segments must be immediately preceded by a slash in '/r/foo:roomId/history'"`
	)
	expect(() => convertReactToVercel('r/:roomId/history')).toThrowErrorMatchingInlineSnapshot(
		`"Route paths must start with a slash, but 'r/:roomId/history' does not"`
	)
})

describe('extractContentPaths', () => {
	it('only includes routes with content', () => {
		expect(extract(<Route path="/foo" element={<div></div>} />)).toEqual(['/foo'])
		expect(extract(<Route path="/foo" Component={() => 'hi'} />)).toEqual(['/foo'])
		expect(
			extract(<Route path="/foo" lazy={() => Promise.resolve({ Component: () => 'foo' })} />)
		).toEqual(['/foo'])
		expect(extract(<Route path="/foo" loader={async () => null} />)).toEqual(['/foo'])
		expect(extract(<Route path="/foo"></Route>)).toEqual([])
		expect(extract(<Route path="/foo"></Route>)).toEqual([])
		expect(
			extract(
				<Route path="/foo">
					<Route path="bar" />
				</Route>
			)
		).toEqual([])
	})
	it('does not include parent routes without content', () => {
		expect(
			extract(
				<Route path="/foo">
					<Route path="bar" element={<div></div>} />
				</Route>
			)
		).toEqual(['/foo/bar'])
	})
	it('does include parent routes with content', () => {
		expect(
			extract(
				<Route path="/foo" element={<div></div>}>
					<Route path="bar" element={<div></div>} />
				</Route>
			)
		).toEqual(['/foo', '/foo/bar'])
	})
})
