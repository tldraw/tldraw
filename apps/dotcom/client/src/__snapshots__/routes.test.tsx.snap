// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the_routes 1`] = `
[
  {
    "reactRouterPattern": "/",
    "vercelRouterPattern": "^//?$",
  },
  {
    "reactRouterPattern": "/f/:fileSlug",
    "vercelRouterPattern": "^/f/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/lf/:fileSlug",
    "vercelRouterPattern": "^/lf/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/new",
    "vercelRouterPattern": "^/new/?$",
  },
  {
    "reactRouterPattern": "/p/:fileSlug",
    "vercelRouterPattern": "^/p/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/preview",
    "vercelRouterPattern": "^/preview/?$",
  },
  {
    "reactRouterPattern": "/q",
    "vercelRouterPattern": "^/q/?$",
  },
  {
    "reactRouterPattern": "/r",
    "vercelRouterPattern": "^/r/?$",
  },
  {
    "reactRouterPattern": "/r/:boardId/history",
    "vercelRouterPattern": "^/r/[^/]*/history/?$",
  },
  {
    "reactRouterPattern": "/r/:boardId/history/:timestamp",
    "vercelRouterPattern": "^/r/[^/]*/history/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/r/:roomId",
    "vercelRouterPattern": "^/r/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/ro/:roomId",
    "vercelRouterPattern": "^/ro/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/s/:roomId",
    "vercelRouterPattern": "^/s/[^/]*/?$",
  },
  {
    "reactRouterPattern": "/ts-side",
    "vercelRouterPattern": "^/ts-side/?$",
  },
  {
    "reactRouterPattern": "/v/:roomId",
    "vercelRouterPattern": "^/v/[^/]*/?$",
  },
]
`;
